File: algorithms/212.Word_Search_II/Word_Search_II.py
class TrieNode:
    # Initialize your data structure here.
    def __init__(self):
        self.is_string = False
        self.leaves = {}

    # Inserts a word into the trie.
    def insert(self, word):
        cur = self
        for c in word:
            if not c in cur.leaves:
                cur.leaves[c] = TrieNode()
            cur = cur.leaves[c]
        cur.is_string = True

class Solution:
    # @param {character[][]} board
    # @param {string[]} words
    # @return {string[]}
    def findWords(self, board, words):
        visited = [[False for j in xrange(len(board[0]))] for i in xrange(len(board))]
        result = {}
        trie = TrieNode()
        for word in words:
            trie.insert(word)
            
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if self.findWordsRecu(board, trie, 0, i, j, visited, [], result):
                    return True
        
        return result.keys()
    
    def findWordsRecu(self, board, trie, cur, i, j, visited, cur_word, result):
        if not trie or i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or visited[i][j]:
            return
        
        if board[i][j] not in trie.leaves:
            return
        
        cur_word.append(board[i][j])
        next_node = trie.leaves[board[i][j]]
        if next_node.is_string:
            result["".join(cur_word)] = True
        
        visited[i][j] = True
        self.findWordsRecu(board, next_node, cur + 1, i + 1, j, visited, cur_word, result)
        self.findWordsRecu(board, next_node, cur + 1, i - 1, j, visited, cur_word, result)
        self.findWordsRecu(board, next_node, cur + 1, i, j + 1, visited, cur_word, result)
        self.findWordsRecu(board, next_node, cur + 1, i, j - 1, visited, cur_word, result)     
        visited[i][j] = False
        cur_word.pop()
        

File: algorithms/212.Word_Search_II/Word_Search_II.cpp
class TrieNode {
 public:
  vector<TrieNode*> next;
  bool isLeaf;
  TrieNode() {
    isLeaf = false;
    next.resize(26, NULL);
  }
};

class Trie {
 public:
  Trie() { root = new TrieNode(); }

  void insert(string word) {
    TrieNode *it = root;
    for(auto & ch : word) {
      if(it->next[ch-'a'] == NULL) { it->next[ch-'a'] = new TrieNode(); }
      it = it->next[ch-'a'];
    }
    it->isLeaf = true;
  }

  TrieNode *getRoot() { return root; }

 public:
  TrieNode *root;
};

class Solution {

 public:
  vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
    for(auto & word : words) trie.insert(word);
    TrieNode *root = trie.getRoot();
    vector<string> result; if(board.size() == 0) return result;
    string tmp;
    unordered_set<string> resultSet;
    for(int i = 0; i < board.size(); ++i) {
      for(int j = 0; j < board[0].size(); ++j) {
        bt(board, root, i, j, tmp, resultSet);
      }
    }
    for(auto & word : resultSet) result.push_back(word);
    return result;
  }

  void bt(vector<vector<char>>& board, TrieNode *root, int x, int y, string tmp, unordered_set<string> & S) {
    int m = board.size(), n = board[0].size();
    if(x < 0 || x >= m || y < 0 || y >= n || board[x][y] == '*') return;
    char ch = board[x][y];
    if(root->next[ch-'a'] != NULL) {
      root = root->next[ch-'a'];
      if(root->isLeaf) {
        S.insert(tmp + ch);
        //root->isLeaf = false;
      }
      board[x][y] = '*';
      bt(board, root, x + 1, y, tmp + ch, S);
      bt(board, root, x - 1, y, tmp + ch, S);
      bt(board, root, x, y + 1, tmp + ch, S);
      bt(board, root, x, y - 1, tmp + ch, S);
      board[x][y] = ch;
    }
  }

 private:
  Trie trie;
};